### ISSUE TITLE
Must Managed Backing Fields be Public?

### ISSUE THREAD

**Issue by JasonBock** [MarimerLLC/cslaforum#84]
I thought that those `public static readonly PropertyInfo<SomeType> ...` fields had to be public, but @rockfordlhotka isn't sure. He'll check to see if that's the case in UWP, if so, I could add an analyzer to ensure they are always public.

Along those lines....I could also add an analyzer to ensure that they must be static (if you make them instance code compiles but I believe it'll fail at runtime). Also, should they always be `readonly`?

**Comment by ajj3085** [comment]
I thought they had to be public for runtimes that restricted reflection to public members only (like Silverlight, and maybe WP).  If you weren't using such runtimes they didn't need to be public.  We've decided to always make them public though because we may end up on a runtime that would require them to be, so rather than having to make changes when that happens, just make it public now and forget about.

Wouldn't they have to be read only though, since aren't they used behind the scenes as keys into dictionaries (like FieldDataManager)?  Changing the reference later would break things badly, I'd expect.

I like this idea for an analyzer; also, if it could check that they are in-fact static as well, that would be great, since I've forgotten that a few times (which leads to chaos).

**Comment by JasonBock** [comment]
@ajj3085 yeah, @rockfordlhotka was going to look into the UWP side of things to see if they're still required to be `public`. But they must be `static`. That should be an error. Making them `readonly` @rockfordlhotka and I thought that should be a warning, but you make a good point about the dictionary. Maybe that should be an error as well.

**Comment by ajj3085** [comment]
@JasonBock  I was just going off memory, so I did a quick test to see what happens if you change the PropertyInfo<T> field after its initially registered.  I tried both registering a new value for PropertyInfo, and replacing the value with one from another BO.  Both raised exceptions at runtime ("Cannot register property after class instantiated" and "Property was not registered"), so an error might be better; one less thing to let people shoot themselves in the foot with. :-)
